### ISSUE TITLE
Problem with taxon in UKBMS taxon list

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1498]
A verifier has reported a problem whereby records for Dingy Skipper are being displayed with the subspecies name "Erynnis tages tages". They are redetermining these to the binomial "Erynnis tages" but the name continues to be displayed as the trinomial. However, it looks like the name originally entered was the binomial, and it is simply being displayed incorrectly.

This appears to apply only to records from the UKBMS website. An example of a record that has been redetermined is here:
https://irecord.org.uk/record-details?occurrence_id=30553915

The record page displays "Erynnis tages", the Explore and verification pages display "Erynnis tages tages".

UKBMS uses taxon list ID = 51. The only anomaly that I can find in the warehouse is that there are two entries for Dingy Skipper, one for the English names and one for the binomial, but both are flagged as rank = subspecies:
- http://warehouse1.indicia.org.uk/index.php/taxa_taxon_list/edit/283032
- https://warehouse1.indicia.org.uk/index.php/taxa_taxon_list/edit/283033

So it looks like these two taxa should be edited to set them to rank = species.

Is that the correct thing to do? And will it resolve the problem of the binomial being displayed as a trinomial?

**Comment by johnvanbreda** [comment]
The problem is not just that the rank is wrong, but in particular that the search code is wrong - for entry 283032 on the warehouse as this is the preferred name you would expect the external key to be the same as the search code but they are different. NHMSYS0000550499 is the correct TVK for E. tages (see https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0000550499) but the search code has NBNSYS0000042368 which is for the sub-species (https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NBNSYS0000042368). That causes the Elasticsearch population to lookup the subspecies taxon information when it populates the record.

BTW The 2 taxa on list 51 that you've found are both part of the same concept - I think you can only get to http://warehouse1.indicia.org.uk/index.php/taxa_taxon_list/edit/283032 via the UI and you probably constructed the other link yourself? The 283033 link is just to the common name entry for this taxon. So, you only need to fix 283032 by correcting both the rank and the search code to NBNSYS0000042368.

I've checked for other cases of the same problem with this query:

```sql
select cttl.id as id_with_problem, cttl.search_code, cttl.external_key, cttl.taxon, cttl2.taxon as taxon_mapped_to
from cache_taxa_Taxon_lists cttl 
join cache_taxa_Taxon_lists cttl2 on cttl2.search_code=cttl.search_code and cttl2.taxon_list_id=15
where cttl.taxon_list_id=51 and cttl.preferred=true and cttl.search_code<>cttl.external_key
```

This does find 12 problems, but they all map to ill-formed versions of the same name, so the problem isn't noticeable.
```
id_with_problem	search_code	external_key	taxon	taxon_mapped_to
283068	"NBNSYS0000005535"	"NHMSYS0000516450"	"Lycaena dispar"	"Lycaena dispar"
283098	"NBNSYS0000005557"	"NHMSYS0000516575"	"Nymphalis polychloros"	"Nymphalis polychloros"
283110	"NHMSYS0000516206"	"NHMSYS0021143558"	"Argynnis adippe"	"Argynnis adippe"
283082	"NHMSYS0020755526"	"NHMSYS0021143616"	"Polyommatus (Lysandra) bellargus"	"Polyommatus (Lysandra) bellargus"
283074	"NHMSYS0000501397"	"NHMSYS0021143605"	"Aricia agestis"	"Aricia agestis"
283080	"NHMSYS0020755527"	"NHMSYS0020825965"	"Polyommatus (Lysandra) coridon"	"Polyommatus (Lysandra) coridon"
283102	"NHMSYS0000502940"	"NHMSYS0021143568"	"Inachis io"	"Inachis io"
283106	"NHMSYS0000516240"	"NHMSYS0021143547"	"Boloria selene"	"Boloria selene"
283112	"NHMSYS0000516211"	"NHMSYS0021297309"	"Argynnis aglaja"	"Argynnis aglaja"
283032	"NBNSYS0000042368"	"NHMSYS0000550499"	"Erynnis tages"	"Erynnis tages subsp. tages"
283030	"NHMSYS0020755521"	"NHMSYS0021143464"	"Ochlodes sylvanus"	"Ochlodes sylvanus"
283110	"NHMSYS0000516206"	"NHMSYS0021143558"	"Argynnis adippe"	"Argynnis adippe"
283042	"NBNSYS0000005518"	"NHMSYS0021143487"	"Colias alfacariensis"	"Colias alfacariensis"
```

**Comment by kitenetter** [comment]
@johnvanbreda thanks for the explanation. For the taxa listed as having an "id_with_problem", can I go ahead and edit the search codes directly in the warehouse UI? And then for Erynnis tages I will also edit the rank to "species".

**Comment by johnvanbreda** [comment]
Yes, that should be fine - afterwards I can run a query to update the documents in Elasticsearch.

**Comment by kitenetter** [comment]
@johnvanbreda I have updated the taxa where the name strings are the same, but I'm not sure what to do with the ones that have changed:
- Argynnis adippe is now Fabriciana adippe https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0021459983
- Polyommatus (Lysandra) bellargus is now Polyommatus bellargus https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0021143616
- Polyommatus (Lysandra) coridon is now Polyommatus coridon https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0020825965
- Inachis io is now Aglais io https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0021143568
- Argynnis aglaja is now Speyeria aglaja https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0021297309

Can you let me know which of these should store the TVK for the name string in question, and which is for the TVK of the current preferred name (where the name string is a synonym):
- "Accepted name unique identifier (external key)"
- "Taxon name unique identifier (search code)"

It looks like the first of those is the current preferred name and the second is the actual name, but I'm doubting myself.

For the taxa where the name in UKBMS maps to a different preferred name in UKSI, should I use the old name TVK as the "Taxon name unique identifier (search code)", and the preferred name TVK as the "Accepted name unique identifier (external key)"?

Which means that the UKBMS taxon list will continue to store the old names. There is an argument for updating to the new names but I don't know if I should edit the name strings or not.

Finally, you asked if I'd constructed the link for the English name "Dingy Skipper", but that URL does exist in the warehouse:
https://warehouse1.indicia.org.uk/index.php/taxa_taxon_list/edit/283033

It is given the rank of "Subspecies" which is not correct, and it has no "Taxon name unique identifier (search code)" - I'm not sure if it needs one. So I'm unclear if I need to edit the English name TVKs and/or ranks to match the scientific names that I've edited.

**Comment by robin-hutchinson** [comment]
Running the butterfly recorder feedback emails, @simonrolph found that Peacock from UKBMS is exporting as Inachis io. We've found a work around to get the UKSI preferred names (same as Martin) between the two systems (as the external key is correct):

select cttl.preferred_taxon, cttl2.preferred_taxon  as preferred_taxon_from_uksi, cttl.external_key
from cache_taxa_taxon_lists cttl
join cache_taxa_taxon_lists cttl2 on cttl.external_key  = cttl2.search_code
where cttl.taxon_list_id = 51
and cttl2.taxon_list_id = 15
and cttl2.preferred_taxon != cttl.preferred_taxon
and cttl.preferred  = TRUE

So we can match back, but this might mean they are still appearing incorrectly in downloads (as the incorrect name currently appears through the API):

preferred_taxon | preferred_taxon_from_uksi | external_key
-- | -- | --
Melitaea   (Mellicta) athalia | Melitaea athalia | NHMSYS0000503190
Vanessa (Cynthia) cardui | Vanessa   cardui | NBNSYS0100005950
Neozephrus   (Quercusia) quercus | Favonius quercus | NHMSYS0021002801
Inachis io | Aglais   io | NHMSYS0021143568
Argynnis   adippe | Fabriciana adippe | NHMSYS0021143558
Small/Essex Skipper | Thymelicus   lineola/sylvestris | NHMSYS0020784146
Plebeius   (Plebejus) argus | Plebejus argus | NBNSYS0000005539
Euphydryas (Eurodryas) aurinia | Euphydryas   aurinia | NHMSYS0000516340
Polyommatus   (Lysandra) bellargus | Polyommatus bellargus | NHMSYS0021143616
Polyommatus (Lysandra) coridon | Polyommatus   coridon | NHMSYS0020825965
Argynnis   adippe | Fabriciana adippe | NHMSYS0021143558
Argynnis aglaja | Speyeria aglaja | NHMSYS0021297309
